packages/flutter_tools/lib/src/flutter_plugins.dart

```diff
@@ -1157,127 +1157,242 @@ List<PluginInterfaceResolution> resolvePlatformImplementation(
+  final bool hasInlineDartImpl = _hasPluginInlineDartImpl(plugin, platformKey);
+
+  if (hasInlineDartImpl) {
+    final String? implementsPackage = plugin.implementsPackage;
+
+    // Only can serve, if the plugin has a dart inline implementation
+    if (implementsPackage != null && implementsPackage.isNotEmpty) {
+      return implementsPackage;
+    }
+
+    if (_isEligibleDartSelfImpl(plugin, platformKey)) {
+      // The inline Dart plugin implements itself.
+      return plugin.name;
+    }
+  }
+
+  return null;
+}
+
+/// Determine if this [plugin] (or package) references a default plugin with an
+/// implementation for the given platform [platformKey].
+/// If so, return the plugin name, which provides the default implementation.
+///
+/// Options:
+///   * The [plugin] (e.g. 'url_launcher') references a default implementation
+///     (e.g. 'url_launcher_linux').
+///   * The [plugin] (e.g. 'url_launcher') implements itself and then also
+///     serves as its own default implementation.
+///   * The [plugin] does not reference a default implementation.
+String? _getDefaultImplPlugin(Plugin plugin, String platformKey) {
+  final String? defaultImplPluginName =
+      plugin.defaultPackagePlatforms[platformKey];
+  if (defaultImplPluginName != null) {
+    return defaultImplPluginName;
+  }
+
+  if (_hasPluginInlineDartImpl(plugin, platformKey) &&
+      _isEligibleDartSelfImpl(plugin, platformKey)) {
+    // The inline Dart plugin serves as its own default implementation.
+    return plugin.name;
+  }
+
+  return null;
+}
+
+/// Determine if the [plugin]'s inline dart implementation for the
+/// [platformKey] is eligible to serve as its own default.
+///
+/// An app-facing package (i.e., one with no 'implements') with an
+/// inline implementation should be its own default implementation.
+/// Desktop platforms originally did not work that way, and enabling
+/// it unconditionally would break existing published plugins, so
+/// only treat it as such if either:
+/// - the platform is not desktop, or
+/// - the plugin requires at least Flutter 2.11 (when this opt-in logic
+///   was added), so that existing plugins continue to work.
+/// See https://github.com/flutter/flutter/issues/87862 for details.
+bool _isEligibleDartSelfImpl(Plugin plugin, String platformKey) {
+  final bool isDesktop = platformKey == 'linux' || platformKey == 'macos' || platformKey == 'windows';
+  final semver.VersionConstraint? flutterConstraint = plugin.flutterConstraint;
+  final semver.Version? minFlutterVersion = flutterConstraint != null &&
+      flutterConstraint is semver.VersionRange ? flutterConstraint.min : null;
+  final bool hasMinVersionForImplementsRequirement = minFlutterVersion != null &&
+      minFlutterVersion.compareTo(semver.Version(2, 11, 0)) >= 0;
+  return !isDesktop || hasMinVersionForImplementsRequirement;
+}
+
+/// Determine, if the plugin provides an inline dart implementation.
```

Comment:
Nit: remove the comma here.

Reply:
Done in 724a0daa7b07636f88019e7a4c41d492c356b458